python/examples/core/nevergrad_parallel_utils.py

```diff
@@ -351,120 +206,145 @@ def finalize_parallel_search(scheduler: NGSchedulerInterface, \
-  done = False
-  while not done:
-    time.sleep(sleep_time)
-    for idx in range(len(tasks)):
-      # Invariant: Always decrease the timer for this process.
-      task = tasks[idx]
-      if task is None:
-        continue
-
-      process = task.process
-      # This process finished by itself, join with it and mark it joined.
-      if not process.is_alive():
-        process.join()
-        task.joined_with_root = True
-        done = True
-        continue
-
-      # This process needs to continue.
-
-    if done:
-      break
-
-  return [
-      idx for idx, e in enumerate(tasks)
-      if e is not None and e.joined_with_root == True
-  ]
-
-
-################################################################################
-### Multiprocess optimization loops.
+### Multiprocess optimization loop.
 ################################################################################
 
 def async_optim_loop(problem_definition: ProblemDefinition, \
                      scheduler: NGSchedulerInterface,
                      optimizer,
                      parsed_args):
-  """Asynchronous NG scheduling problem with multi-process evaluation of proposals.
-  """
+  """Asynchronous NG scheduling with multi-process evaluation of proposals."""
+
+  # Helper variables.
+  num_cpus_per_benchmark = parsed_args.num_cpus_per_benchmark
+
+  # A benchmark may run on multiple CPUs. Must divide cpu_count evenly.
+  assert cpu_count() % parsed_args.num_cpus_per_benchmark == 0, \
+    f'num_cpus_per_benchmark: {parsed_args.num_cpus_per_benchmark} ' + \
+    f'must divide the number of cpus: {cpu_count()}'
+  num_concurrent_benchmarks = int(cpu_count() / num_cpus_per_benchmark)
+
+  # Reader-writer locks are used to ensure that nothing else in running on a
+  # CPU range while a benchmark is running. If no benchmark is running, multiple
+  # compilations may be running on a CPU range (even more than cores available).
+  mp_manager = mp.Manager()
+  benchmark_locks = [RWLock() for idx in range(num_concurrent_benchmarks)]
+
+  # Concurrent queue for benchmark results.
+  results_queue = mp_manager.Queue()
+
+  # Concurrent queue for available CPUs. Every mp.Pool process is pinned to one
+  # CPU from the queue. During the benchmark phase, it is pinned to a CPU range.
+  available_cpus_queue = mp_manager.Queue()
+  for i in range(parsed_args.num_parallel_tasks):
+    available_cpus_queue.put(i % cpu_count())
+
+  shutdown_event = mp_manager.Event()
+
+  def init_proccess(available_cpus_queue_arg,
+                    parsed_args_arg,
+                    results_queue_arg,
+                    rw_locks_arg: tp.Sequence[RWLock],
+                    shutdown_event_arg):
+    """Initialize an mp.Pool process.
+
+    Store frequently used objects in global variables and select a pinned
+    compilation CPU for this process.
+    """
+
+    # Store arguments in global variables.
+    global available_cpus_queue, parsed_args, results_queue, rw_locks, shutdown_event
```

Comment:
eeesh .. do we really need this ?

Reply:
I could put them into a single objects, so that we only have one global variable.